src/net.cpp

```diff
@@ -340,71 +340,161 @@ bool CConnman::CheckIncomingNonce(uint64_t nonce)
+        // name catch this early.
+        LOCK(cs_vNodes);
+        CNode* pnode = FindNode((CService)addrDest);
+        if (pnode)
+        {
+            pnode->MaybeSetAddrName(std::string(pszDest));
+            LogPrintf("Failed to open new connection, already connected\n");
+            return false;
+        }
+    }
 
-        // Add node
-        NodeId id = GetNewNodeId();
-        uint64_t nonce = GetDeterministicRandomizer(RANDOMIZER_ID_LOCALHOSTNONCE).Write(id).Finalize();
-        CNode* pnode = new CNode(id, nLocalServices, GetBestHeight(), hSocket, addrConnect, CalculateKeyedNetGroup(addrConnect), nonce, pszDest ? pszDest : "", false);
-        pnode->nServicesExpected = ServiceFlags(addrConnect.nServices & nRelevantServices);
-        pnode->AddRef();
+    // Configure proxy, if any. At this point, addrDest is resolved and valid unless we're connecting through
+    // a proxy which will be handling the resolve.
+    proxyType proxy;
+    if (addrDest.IsValid()) {
+        GetProxy(addrDest.GetNetwork(), proxy);
+    } else if (!GetNameProxy(proxy)) {
+        return false;
+    }
+
+    // Create socket
+    SOCKET hSocket = INVALID_SOCKET;
+    if (proxy.IsValid()) {
+        if (!CreateSocket(proxy.proxy, hSocket)) {
+            return false;
+        }
+    } else {
+        if (!CreateSocket(addrDest, hSocket)) {
+            return false;
+        }
+    }
+    if (!IsSelectableSocket(hSocket)) {
+        LogPrintf("Cannot create connection: non-selectable socket created (fd >= FD_SETSIZE ?)\n");
+        CloseSocket(hSocket);
+        return false;
+    }
 
-        return pnode;
-    } else if (!proxyConnectionFailed) {
-        // If connecting to the node failed, and failure is not caused by a problem connecting to
-        // the proxy, mark this as an attempt.
-        addrman.Attempt(addrConnect, fCountFailure);
+    // Connect
+    bool connected = false;
+    if (proxy.IsValid()) {
+        if (!ConnectSocketDirectly(proxy.proxy, hSocket, nConnectTimeout)) {
+            // This does not count as an addrman attempt, since the address was never actually tried.
+            return false;
+        }
+        ProxyCredentials random_auth;
+        const ProxyCredentials* proxy_auth = nullptr;
+        if (proxy.randomize_credentials) {
+            static std::atomic_int counter;
+            random_auth.username = random_auth.password = strprintf("%i", counter++);
+            proxy_auth = &random_auth;
+        }
+        if (addrDest.IsValid()) {
+            strDest = addrDest.ToStringIP();
+            port = addrDest.GetPort();
+        }
+        connected = Socks5(strDest, (unsigned short)port, proxy_auth, hSocket);
+    } else {
+        connected = ConnectSocketDirectly(addrDest, hSocket, nConnectTimeout);
     }
 
-    return NULL;
+    addrman.Attempt(addrDest, fCountFailure);
+
+    if (!connected) {
+        return false;
```

Comment:
Not strictly your fault, but I think you're exposing a bug in Socks5 here - Socks5 fails to CloseSocket() prior to returning false in one place (Proxy username or password too long), so I think it'd be better to just CloseSocket here instead of further down to make it harder to miss such things.